[GSB] Reduce use of potential archetypes when describing constraints

DougGregor opened the pull request:
Start delaying the realization of potential archetypes in a few ways:

* `Constraint` can now store a dependent type that will be lazily (and explicitly) realized to a potential archetype
* The `RequirementSource` APIs traffic mostly in dependent types now

DougGregor committed:
[GSB] Abstract away the PotentialArchetype stored in a constraint.

DougGregor committed:
[GSB] Switch Constraint's "subject" over to an unresolved type.

We're still eagerly resolving the subject to a potential archetype
everywhere, but this refactoring allows us to start introducing laziness.

DougGregor committed:
[GSB] Move definition of GenericSignatureBuilder::Constraint out of the class.

It's getting big, and there will soon be an ordering dependency with
FloatingRequirementSource. NFC

DougGregor committed:
[GSB] Start moving RequirementSource off of potential archetypes.

Eliminate potential archetypes from most of the public interface of
RequirementSource.

DougGregor committed:
[GSB] Eliminate potential archetypes from minimizing requirement sources.

Refactor the interfaces that involve “walking” a requirement source
(e.g., minimization, computation of the affected type, etc.) to rely
only on interface types and not potential archetypes.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[GSB] Eliminate potential archetypes from requirement sources entirely.

DougGregor committed:
[GSB] Teach expandConformanceRequirement() to use a resolved type.

Eliminates another too-early source of realization of potential archetypes.

DougGregor committed:
[GSB] Eliminate a debugging-only source of potential archetype realization.

slavapestov commented on include/swift/AST/GenericSignatureBuilder.h:
```diff
@@ -1546,17 +1546,16 @@ class GenericSignatureBuilder::PotentialArchetype {
   /// Retrieve the generic signature builder with which this archetype is
   /// associated.
-  /// FIXME: Only EquivalenceClassVizIterator gets to use this.
   GenericSignatureBuilder *getBuilder() const {
```
So is this still needed?

DougGregor replied:
There are several callers still, but not so many that we can't eliminate their dependence on `getBuilder()`.

DougGregor replied:
*sigh*, that took longer than expected. It's gone now.

DougGregor committed:
[GSB] Eliminate most uses of PotentialArchetype::getBuilder().

The remaining two uses after this refactor are... stubborn.

DougGregor committed:
[GSB] Eliminate the penultimate caller to PotentialArchetype::getBuilder().

DougGregor committed:
[GSB] Plumb generic params through getDependentType() fully.

This allows us to eliminate `getBuilder()`.

DougGregor committed:
[GSB] Eliminate the GenericSignatureBuilder from PotentialArchetype.

Now that getBuilder() is gone, stash an ASTContext into root
PotentialArchetypes instead of a GenericSignatureBuilder. This eliminates
some ickiness where we had to re-root potential archetypes when moving a
GenericSignatureBuilder.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
@graydon there was a failure with the compiler performance test

DougGregor commented:
@swift-ci please smoke test compiler performance

DougGregor commented:
@swift-ci please smoke test compiler performance
